### DISCUSSION TITLE
ANNOUNCEMENT: Logica **is** switching to DuckDB by default

### DISCUSSION THREAD

**Discussion by EvgSkv** [EvgSkv/logica#409] [General]
Given that most new users don't have Google Cloud project handy, it's best for adoption to switch to run on SQLite or DuckDB by default.

SQLite is great, but it can have very hard to understand unexpected behavior for complex aggregation. Thus I am planning to use DuckDB by default. DuckDB needs installation, but it's a zero depenendency library and is a bliss to install.

BigQuery remains fully supported and the easiest way to run Logica on thousands of machines. The switch would be happening to ease adoption.

If you have any concerns please reply in this thread! 

Tentative plan is to switch in January 2025.

Thank you!

**Comment by EvgSkv** [comment]
UPDATE: I have heard no complaints, so the plan is determined. I'll update title.

But if you have concerns - please reach out at any moment.

**Reply by turbo** [reply]
Could you go into a few more details about this:

> SQLite is great, but it can have very hard to understand unexpected behavior for complex aggregation

**Reply by EvgSkv** [reply]
Yeah, this is a good question. I was working a lot with SQLite as Logica engine before postgres and duckdb support was implemented, but I wasn't careful to "keep the evidence". SQLite is great and errors occur only in some complex situations, but this also makes them hard to reproduce. Thus unfortunately I don't have an example at hand. 

Yet, errors were not the only factor why DuckDB was chosen as default. Here is some elaboration about the differences between SQLite and DuckDB that I was thinking about. Luckily engine is easy to change and SQLite or DuckDB can be chosen on a case by case by the user.

Given the speed and robustness of DuckDB it seems to me that it is a better _default_ experience for new users of Logica.
Need to intstall is not much of a burden since DuckDB is extremely careful about its own dependencies.

Strong sides of SQLite in Logica:
  * Needs no installation.
  * Native access to the data in popular format.
  * Handles dynamic typing.

Challenges of working with SQLite in Logica:
  * Misinterprents some complex nested aggregation expressions.
  * JSON errors are hard to understand (this could be solved by improving Logica compilation)

Strong sides of DuckDB:
  * Very fast.
  * Robust execution for aggregation of arbitrary complexity.
  * Reads data in many formats, including CSVs from internet.

**Comment by EvgSkv** [comment]
UPDATE: Main repo is switched to use DuckDB by default. It will be there for a few weeks or months to catch possible bugs and in the next pip realease will contain the change.
